pontoon/base/models.py

```diff
@@ -284,53 +284,82 @@ def latest_activity(self, locale=None):
+        qs = (
+            Stats.objects
+            .select_related('resource')
+            .filter(resource__project=self, resource__entity__obsolete=False)
+            .distinct()
+        )
+        locales = self.locales.prefetch_related(
+            Prefetch(
+                'stats_set',
+                queryset=qs,
+                to_attr='fetched_stats'
+            )
+        )
+
+        # Project Pages with prefetched Resources
+        pages = self.subpage_set.prefetch_related(
+            'resources'
+        )
+
+        has_pages = pages.exists()
+
+        for locale in locales:
             locale_details = []
 
-            # Get subpage name and stats if subpages defined
-            if pages.exists():
+            # If project has defined subpages, return their names with
+            # corresponding project stats. If subpages have defined resources,
+            # only include stats for page resources.
+            if has_pages:
 
                 # List only subpages, whose resources are available for locale
-                pages = pages.filter(resources__stats=stats) or pages
+                pages = pages.filter(resources__stats__in=locale.fetched_stats) or pages
 
                 for page in pages:
+                    entity_count = translated_count = approved_count = 0
+                    paths = page.resources.values_list('path')
+                    page_stats = locale.fetched_stats[:]  # Clone
 
-                    # Stats for subpage resources only if defined
+                    # Only include stats for page resources (if defined)
                     if page.resources.exists():
-                        page_stats = stats.filter(resource__in=page.resources.all())
-
-                    # Or for the entire project
-                    else:
-                        page_stats = stats
-
-                    dictionary = {
-                        'resource__path': page.name
-                    }
-                    dictionary.update(page_stats.aggregate(
-                        resource__entity_count=Sum('resource__entity_count'),
-                        translated_count=Sum('translated_count'),
-                        approved_count=Sum('approved_count')
-                    ))
-                    locale_details.append(dictionary)
-
-            # Get resource paths and stats if more than one resource present
-            elif resources.count() > 1:
-                locale_details = stats.order_by('resource__path').values(
-                    'resource__path',
-                    'resource__entity_count',
-                    'translated_count',
-                    'approved_count',
-                )
-
-            details[locale.code.lower()] = list(locale_details)
+                        for stats in locale.fetched_stats:
+                            if stats.resource.path not in [x[0] for x in paths]:
+                                page_stats.remove(stats)
```

Comment:
You clone the list and remove the stats that don't match, but then below you check again if the path is in the list anyway. I think that the cloning and removal can just be removed, and you can keep the `stats.resource.path in paths` check in the for loop below to keep it simple.

I suspect, based on the check for `if page.resources.exists()`, that you only want to do this filtering sometimes, but the way it's written now I think it will always do the filtering. Maybe that's what needs to change?